Fix #3633 revision 2

deroko opened the pull request:
This fixes issue #3633 by isolating `PEB->Fls*` to global variable, thus Windows 1903 changes in PEB will not affect FLS isolation.

derekbruening commented on core/win32/drwinapi/kernel32_proc.c:
```diff
@@ -48,8 +48,14 @@ void
+    /* FIXME: i#3633: Fix for Windows 10 1903 where PEB->Fls* is replaced with
+     * PEB->SparePointers and PEB->SpareUlongs. Maybe this code shouldn't even
+     * check for peb->FlsBitmap, as in ntdll_redir.c we are anyhow allocating
+     * own FLS structure globaly and we don't relay at all on Windows FLS
```
nit: spelling: s/relay/rely/

derekbruening commented on core/win32/drwinapi/kernel32_proc.c:
```diff
@@ -48,8 +48,14 @@ void
+     * PEB->SparePointers and PEB->SpareUlongs. Maybe this code shouldn't even
+     * check for peb->FlsBitmap, as in ntdll_redir.c we are anyhow allocating
+     * own FLS structure globaly and we don't relay at all on Windows FLS
+     * implementation
```
nit: style violation: missing final `.`

derekbruening commented on core/win32/drwinapi/ntdll_redir.c:
```diff
@@ -833,6 +833,23 @@ redirect_RtlPcToFileHeader(__in PVOID PcValue, __out PVOID *BaseOfImage)
 }
+/* XXX i#3633: Isolate PEB->Fls* for Windows 10 1903 where Fls is stored inside of
+ * ntdll.dll internal variables. This code isolates PEB->Fls* in dynamorio global
+ * variable, thus we don't use PEB anymore for FLS handling
```
nit: style violation: missing final `.`

derekbruening commented on core/win32/drwinapi/kernel32_proc.c:
```diff
@@ -48,8 +48,14 @@ void
 kernel32_redir_init_proc(void)
 {
     PEB *peb = get_own_peb();
+    /* FIXME: i#3633: Fix for Windows 10 1903 where PEB->Fls* is replaced with
```
nit: style: no colon: `s/FIXME: i/FIXME i/`

derekbruening commented on core/win32/drwinapi/ntdll_redir.c:
```diff
@@ -833,6 +833,23 @@ redirect_RtlPcToFileHeader(__in PVOID PcValue, __out PVOID *BaseOfImage)
+    ULONG FlsBitmapBits[5];
+} FLS_GLOB_DATA, *PFLS_GLOB_DATA;
+
+PFLS_GLOB_DATA redirFls;
```
name style violation: `redir_fls`

derekbruening commented on core/win32/drwinapi/ntdll_redir.c:
```diff
@@ -843,38 +860,33 @@ ntdll_redir_fls_init(PEB *app_peb, PEB *private_peb)
-    /* We initialize this to zero (required for redirect_RtlProcessFlsData),
-     * and we assume no pre-existing entries (just like we start FlsList* empty).
+    /* XXX: i#3633: Isolate PEB->Fls* for Windows 10 1903 and previous version of Windows
+     * thus same code can be used on pre Windows 10 1903 where FLS was held in PEB
```
nit: style: missing final `.`

derekbruening commented on core/win32/drwinapi/ntdll_redir.c:
```diff
@@ -833,6 +833,23 @@ redirect_RtlPcToFileHeader(__in PVOID PcValue, __out PVOID *BaseOfImage)
+    LIST_ENTRY FlsListHead;
+    PVOID *FlsCallback;
+    ULONG FlsHighIndex;
+    ULONG FlsBitmapBits[5];
```
Our ntdll.h lists this as `DWORD FlsBitmapBits[4]`.  Why 5 instead of 4 here?  Is this extended to 5 in 1903?  Worth a comment.

deroko replied:
Correct, it should be 4. I've made type error. However, Windows 10 1903 apparently allows up to 0xFF0 FLS slots, according to this code from ntdll.dll:

```
.text:4B305F63                 cmp     eax, 0FF0h
.text:4B305F68                 jnb     loc_4B305FEF
```

compared to the old code 

```
.text:0DCE7CF8                 cmp     eax, 7Eh ; '~'
.text:0DCE7CFB                 ja      loc_DD0B617
```

derekbruening commented on core/win32/drwinapi/ntdll_redir.c:
```diff
@@ -843,38 +860,33 @@ ntdll_redir_fls_init(PEB *app_peb, PEB *private_peb)
-    HEAP_TYPE_FREE(GLOBAL_DCONTEXT, private_peb->FlsBitmap, RTL_BITMAP, ACCT_LIBDUP,
+    HEAP_ARRAY_FREE(GLOBAL_DCONTEXT, redirFls->FlsCallback, PVOID,
+                    redirFls->FlsBitmap->SizeOfBitMap, ACCT_LIBDUP, UNPROTECTED);
+    HEAP_TYPE_FREE(GLOBAL_DCONTEXT, redirFls->FlsBitmap, RTL_BITMAP, ACCT_LIBDUP,
```
Missing a free of redirFls itself.

derekbruening commented on core/win32/drwinapi/kernel32_proc.c:
```diff
@@ -48,8 +48,14 @@ void
+     * check for peb->FlsBitmap, as in ntdll_redir.c we are anyhow allocating
+     * own FLS structure globaly and we don't relay at all on Windows FLS
+     * implementation
+     */
```
While this does give us a single solution independent of the Windows version, it does mean that if there's some code (say in a Windows system library we've loaded privately for the client) that examines PEB->Fls* we'll hit problems: i.e., we're relying on intercepting *everything* that touches Fls, whereas un-intercepted code has a chance of working if the data was in the same place.  This approach still seems fine since we are intercepting for other reasons, but it may be worth a comment acknowledging the assumption on interception.

deroko replied:
Comment can be reformulated to fully describe what is (and why) going on.

What bothers me here, is certain scenario which I want to test. If somebody knows answer to next situation would be nice to jump in and spare time in debugging. 

Here is scenario which puzzles me (but I have to verify):
1. lets load ucrtbase.dll from client - private_load. flsIndex will be allocated with redirection, and stored in global var in ucrtbase.dll as a part of ucrtbase!CRT_INIT. 
2. teb->FlsData at this point is changed to private FlsData (priv_fls_data).
3. teb->FlsData[flsIndex] = something_from_ucrtbase.dll
4. Now, maybe through delay load import, this dll is being resolved in original app and API gets called which uses global flsIndex. In this case teb->FlsData (app_fls_data) is original from the app, and crt expects to find valid data there. It should either get wrong data (NULL) or FLS set by some other dll. Wanted/expected data resided inside of a dcontext->client_data->priv_fls_data, and not inside of a dcontext->client_data->app_fls_data;

I will run this test, and post update. Just to be 100% sure that code doesn't run into some weird case which could lead to hard-to-debug crashes.

derekbruening replied:
I don't follow: if the app loads ucrtbase, ucrtbase!CRT_INIT should be executed in an app context which will initialize the app's version of the FLS.

deroko replied:
I don't know, maybe or maybe not depending if the app will be using it or not. One solution would be that teb->FlsData is not replaced, but as redir_Fls*/redir_RtlFls* is applied in Get/SetValue to directly use dcontext->client_info->priv_fls_data; as flsindex will always point to real data. Yes, it could happen that client by using these APIs will obtain FLS data in private dynamorio memory, but there will be no conflict using one globally saved flsindex. 

Apparently this is not a big issue, as nobody noticed it over the years. I would keep it as is, and keep it in mind in case somebody ever runs into it.

derekbruening commented on core/win32/drwinapi/ntdll_redir.c:
```diff
@@ -843,38 +860,33 @@ ntdll_redir_fls_init(PEB *app_peb, PEB *private_peb)
 
-    /* We initialize this to zero (required for redirect_RtlProcessFlsData),
-     * and we assume no pre-existing entries (just like we start FlsList* empty).
+    /* XXX: i#3633: Isolate PEB->Fls* for Windows 10 1903 and previous version of Windows
```
nit: no colon: `s/XXX:/XXX/`

AssadHashmi commented:
run arm tests

derekbruening commented:
Note that Appveyor is hitting some failures, including several instances of STATUS_STACK_BUFFER_OVERRUN:
```
[00:10:44] 222/256 Test #223: code_api|tool.drcacheoff.burst_static ........................***Failed    0.25 sec
[00:10:44] Running cmd |C:/projects/dynamorio/build/build_debug-internal-32/clients/bin32/tool.drcacheoff.burst_static.exe|
[00:10:44] CMake Error at C:/projects/dynamorio/suite/tests/runmulti.cmake:106 (message):
[00:10:44]   *** cmd failed (Exit code 0xc0000409
```

deroko commented:
Whew, managed to reproduce test cases locally. Investigating them atm.

deroko commented:
This C0000409 crashes are very interesting, and they come from funny conditions. I think I've nailed it to the right one.

As I can see, we are dealing with crashes in binaries with static dynrio compiled in, and the reason at least on windows 10 1903 is that Fls* apis are not redirected, as they can't be at this early stage when they are resolved (crt!mtinit), and DynamoRIO is being invoked from main() which is much later in the process, thus they share same global vars in ntdll, which will yield crash.

On other windows versions, following logic, similar happens but this time, as PEB->Fls* is isolated crash will not happen. 

What I can do, is to restore PEB isolation (as it was) and let windows 1903 to simulate PEB, which will crash these tests on 1903, but let them pass prior to 1903 as is done in the 1st try at resolving this issue : https://github.com/deroko/dynamorio/commit/42471c381395faf6db2154b7df572990695f4e42

derekbruening commented:
Static clients are inherently fragile, and we can't really support the use of app library routines in static clients b/c as you said we can't isolate or redirect.  But by keeping static clients carefully isolated manually they can work, and these should be isolated -- wait, I see the use of `std::cerr`in burst_static.cpp.  Probably that is the source of the problem.  That should be dr_fprintf: no non-DR library routines should be used in these.

derekbruening commented:
If we do `s/std::cerr/dr_fprintf(STDERR/` does it work, or is there sthg else inside drmemtrace that is running Windows library code?

deroko commented:
Yes, this was the issue, as it was using `write_nolock` from CRT which was using FLS.

derekbruening commented:
This does show a scenario where we can't do full code isolation, but we can do some data isolation: so having both may be a good idea.  It adds code complexity of course with the 1903 split so it could be argued that that complexity outweighs the layered isolation benefit.

deroko commented:
I've nailed it down to maybe proper fix that could be implemented, to fight cases when it even slips, and Fls gets executed with priv_fls_data. 

Here is workflow: 
```
1. teb->priv_fls_data (= NULL)
2. write_nolock (or anything else) calls RtlFlsSetValue
3. since FlsData is NULL, allocation happens and FlsData is added to the list_entry of all Allocated FlsData (drrio mem endsup in internal LIST_ENTRY)
4. shutdown is called and pointer to priv_fls_data remains in LIST_ENTRY, but memory is freed
5. on next run, we have another RtlFlsSetValue which tries to use LIST_ENTRY to add another entry, but since previous priv_fls_data is not valid anymore crash will happen on 1903.
```

with private_peb this wasn't happening, as everything was in private_peb, but now in 1903 this becomes part of internal LIST_ENTRY. What is possible, is to keep private_peb for backward compatibility, and add code that for every thread_exit adds code to do teb->priv_fls_data unlinking. This will make sure that this is removed from internal ntdll variables. My question would be, what is the best place to put this unlinking for every thread?

derekbruening commented:
> My question would be, what is the best place to put this unlinking for every thread?

You mean you'd be adding ntdll_redir_fls_thread_exit() or sthg?  Could have os_loader_thread_exit => drwinapi_thread_exit => ntdll/kernel32_redir_thread_exit(); or direct call from loader like call to ntdll_redir_fls_init() today.

deroko commented:
Yes, that would be one logical solution for 1903 fix, so even if it slips, FLS is unlinked and no crash will happen.

Wow, if that one is possible to use it would be awesome. I thought it is only used by main thread only. I'll do test quickly.

AssadHashmi commented:
run arm tests

deroko commented:
So final fix, I will merge into this branch, still running all tests. I'll keep private_peb just in case that FLS slips on older versions, and on Windows 10 1903 private_peb will be faked, and on thread_exit in ntdll_fls_redir_thread_exit, it will do unlinking on a dconetxt->priv_fls_data. I'll push these changes tomorrow, and hopefully this issue can be closed for now.

deroko commented:
I've pushed new commit. Removed `fls_redir`, and use `private_peb` just in case of old systems. Added `ntdll_redir_fls_thread_exit` to unlink `priv_fls_data`, which makes sure that FLS is removed from internal ntdll.dll in case it endsup there on 1903.

AssadHashmi commented:
run arm tests

derekbruening commented on core/win32/loader.c:
```diff
@@ -412,7 +412,10 @@ os_loader_thread_init_epilogue(dcontext_t *dcontext)
 os_loader_thread_exit(dcontext_t *dcontext)
 {
-    /* do nothing in Windows */
+    /* XXX i#3633: In case of windows 1903 if priv_fls_data endsup in internal lists of
```
Again, I don't think we want the `XXX`.

derekbruening commented on core/win32/drwinapi/ntdll_redir.c:
```diff
@@ -838,15 +838,27 @@ redirect_RtlPcToFileHeader(__in PVOID PcValue, __out PVOID *BaseOfImage)
     /* FLS is supported in WinXP-64 or later */
     ASSERT(get_os_version() >= WINDOWS_VERSION_2003);
+
+    /* FIXME i#3633: Implement FLS isolation in Win10 1903+ where FLS data is no longer
```
We're doing this now, so the word `FIXME` should be removed, right?

derekbruening commented on core/win32/drwinapi/kernel32_proc.c:
```diff
@@ -48,8 +48,11 @@ void
 kernel32_redir_init_proc(void)
 {
     PEB *peb = get_own_peb();
+    /* FIXME i#3633: Implement FLS isolation in Win10 1903+ where FLS data is no longer
```
We have an implementation so no need for the word FIXME, right?  If you want to talk about ways to make the implementation here better feel free, though I would that would be XXX since it's a nice-to-have not a must-have, and the comment should spell out what we'd do differently.

deroko replied:
I could write on this one a bit more, but I at the moment I think it is good to go for 1903. Don't know if DynamoRIO gives ability to intercept APIs from the engine itself - `drwrap_wrap` like functionality. With this, it would be trivial to intercept FLS APIs, and properly redirect them, or handle FLS without ntdll.dll ever touching it. Maybe for the future?

derekbruening replied:
drwrap is for intercepting the app, which is run under DR control (or possibly natively for the not-really-supported Probe API).  Here we need to intercept libraries running natively in tool mode.  The private loader can do this at the inter-library import level, as is done with the Fls functions in drwinapi/ today.

deroko commented on core/win32/drwinapi/drwinapi.h:
```diff
@@ -61,4 +61,10 @@ ntdll_redir_fls_init(PEB *app_peb, PEB *private_peb);
 void
 ntdll_redir_fls_exit(PEB *private_peb);
 
+/* XXX i#3633: Fix Windows 1903 issue. FLS is not held inside of PEB but in private
```
Correct, removed it.

derekbruening commented:
run arm tests

derekbruening commented:
Appveyor hit the type_is_instr assert again: #3320
